Thread Safety Analysis: Support reentrant capabilities

melver opened the pull request:
Introduce the `reentrant_capability` attribute, which may be specified
alongside the `capability(..)` attribute to denote that the defined
capability type is reentrant. Marking a capability as reentrant means
that acquiring the same capability multiple times is safe, and does not
produce warnings on attempted re-acquisition.

The most significant changes required are plumbing to propagate if the
attribute is present to a CapabilityExpr, and introducing ReentrancyCount
to the LockableFactEntry class.

melver committed:
Thread Safety Analysis: Support reentrant capabilities

Introduce the `reentrant_capability` attribute, which may be specified
alongside the `capability(..)` attribute to denote that the defined
capability type is reentrant. Marking a capability as reentrant means
that acquiring the same capability multiple times is safe, and does not
produce warnings on attempted re-acquisition.

The most significant changes required are plumbing to propagate if the
attribute is present to a CapabilityExpr, and then introducing a
ReentrancyCount to FactEntry that can be incremented while a fact
remains in the FactSet.

AaronBallman commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -3990,6 +3990,13 @@ def LocksExcluded : InheritableAttr {
+def ReentrantCapability : InheritableAttr {
+  let Spellings = [Clang<"reentrant_capability">];
+  let Subjects = SubjectList<[Record, TypedefName]>;
+  let Documentation = [Undocumented];
```
We generally don't want new, undocumented attributes. However, I think this is fine because we have dedicated documentation for this elsewhere which is being updated, and this matches the same lack of documentation as the rest of the thread safety attributes.

melver replied:
Ack. Yes, style follows the rest of these attributes.

philnik777 replied:
IMO we could at least add a link to the other documentation. Just saying that it's undocumented in the place where all other attributes are documented is really not helpful.

AaronBallman commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -139,22 +141,41 @@ class FactEntry : public CapabilityExpr {
+    if (kind() != ReenterKind)
+      return false;
+    if (++ReentrancyCount == 0)
+      llvm::report_fatal_error("Maximum reentrancy reached");
```
This should use a real diagnostic rather than reporting a fatal error, though I expect no user will ever run into the diagnostic in practice.

aaronpuchert commented on clang/docs/ThreadSafetyAnalysis.rst:
```diff
@@ -434,6 +434,16 @@ class can be used as a capability.  The string argument specifies the kind of
+capability multiple times is safe.
+
+Note that acquiring the same capability with different access privileges
+(exclusive vs. shared) again is not considered reentrant by the analysis.
```
Perhaps we could add a sentence that this is usually not needed, because the analysis will only complain about double locking that's provably unnecessary. It's only in cases where  locking is happening through abstractions that double locking might be hard to prevent.

melver replied:
Added a Note. PTAL.

aaronpuchert commented on clang/include/clang/Analysis/Analyses/ThreadSafetyCommon.h:
```diff
@@ -271,26 +271,32 @@ class CFGWalker {
-  CapabilityExpr() : CapExpr(nullptr, false) {}
-  CapabilityExpr(const til::SExpr *E, StringRef Kind, bool Neg)
-      : CapExpr(E, Neg), CapKind(Kind) {}
+  static constexpr unsigned FlagNegative = 1u << 0;
+  static constexpr unsigned FlagReentrant = 1u << 1;
```
I wonder if there is a use case for having both. If the capability can be acquired an arbitrary number of times, does it make sense to require it not being held? At least the negative capability should not be required when acquiring a reentrant mutex.

melver replied:
I don't follow. We need a way to query both, so we need at least 2 bits to store that info.

melver replied:
We need 4 states.

I'm adding this test:
```
+class TestNegativeWithReentrantMutex {
+  ReentrantMutex rmu;
+  int a GUARDED_BY(rmu);
+
+public:
+  void baz() EXCLUSIVE_LOCKS_REQUIRED(!rmu) {
+    rmu.Lock();
+    a = 0;
+    rmu.Unlock();
+  }
+};
```

aaronpuchert replied:
That's exactly the question. Does this make sense? If `rmu` can be acquired an arbitrary number of times, why does this require `!rmu`? It can be acquired even if it is already held.

So I guess the question is whether we should not simply disallow `!rmu`.

melver replied:
There might be valid uses for that, even if it's not something we'd expect to be frequent. Developer creativity how to use a feature can sometimes lead to new uses we never expected, and I wouldn't want to prohibit them.

If it's *not wrong semantically*, I tend to not want to impose a certain pattern and imply to the user "you're holding it wrong" even though it's not strictly wrong and they might actually know what they're doing. I've been on the receiving end of such forced restrictions, and it's rather annoying - though quite often it's because the restriction allowed for a simpler implementation, I'd argue in this case that's a moot point.

I think if we were talking about synchronization only, I'd be more inclined to say "this never makes sense", but the feature is now about arbitrary capabilities.

For example, we might want to express that some capability is never held when entering a function for performance reasons, yet that capability itself can be reentrantly acquired on slow paths. I don't see a reason to prohibit uses like that.

aaronpuchert replied:
While I see your point, this is all a bit vague. The purpose of negative capabilities is [documented](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#negative) as helping to prevent double locking, which is not an issue for reentrant mutexes.

> I've been on the receiving end of such forced restrictions, and it's rather annoying

The other side of the coin is that it is very hard to make the analysis more strict after we've released a feature, while making it more permissive is not a problem. Having done a few breaking changes myself, it's something that I'd like to avoid as much as possible.

That's why I tend to err on the side of making the analysis as strict as it can be while still being useful, and if that causes too much friction relax it later.

In this case, the restriction doesn't seem arbitrary at all. The combination simply doesn't make sense by the documented purpose of negative capabilities.

aaronpuchert commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -1831,15 +1852,15 @@ void BuildLockset::handleCall(const Expr *Exp, const NamedDecl *D,
     if (D->hasAttrs() && TagT && Exp->isPRValue()) {
-      std::pair<til::LiteralPtr *, StringRef> Placeholder =
-          Analyzer->SxBuilder.createThisPlaceholder(Exp);
+      auto Placeholder = Analyzer->SxBuilder.createThisPlaceholder(Exp);
```
I'd like to keep an explicit type here for readability. As an alternative, use the typed `std::get`, i.e. `std::get<StringRef>`.

melver replied:
Growing this pair/tuple become a little too complex, so I just made createThisPlaceholder return a CapabilityExpr struct right away.

aaronpuchert reviewed (commented):
I think the biggest issue is that removing `const` from `FactEntry` does not work. You'll have to undo all those changes and instead create a new `FactEntry` for every lock/unlock.

melver commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -163,15 +184,15 @@ using FactID = unsigned short;
 class FactManager {
 private:
-  std::vector<std::unique_ptr<const FactEntry>> Facts;
+  std::vector<std::unique_ptr<FactEntry>> Facts;
```
Fixed.

I guess we have to do clone-mutate-replace of the FactEntries - a little less efficient, but given these structs are relatively small (+ vtables) shouldn't be too bad.

melver commented:
> I think the biggest issue is that removing `const` from `FactEntry` does not work. You'll have to undo all those changes and instead create a new `FactEntry` for every lock/unlock.

Good catch, reworked this.
PTAL.

aaronpuchert reviewed (commented):
Two more things come to mind:
* Consider adding a check to `SemaDeclAttr.cpp` that the new attribute is always accompanied by `capability` or `lockable`. Although I wonder whether that's too early. I'm not sure if we can see the other attributes already.
* You should probably do something in `ThreadSafetyAnalyzer::intersectAndWarn`: if two branches join with different counts, we should warn and drop the fact with higher count.

melver commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -235,6 +266,20 @@ class FactSet {
     return false;
   }
 
+  std::optional<FactID> replaceLock(FactManager &FM, iterator It,
+                                    std::unique_ptr<FactEntry> Entry) {
+    if (It == end())
+      return std::nullopt;
+    FactID F = FM.newFact(std::move(Entry));
+    *It = F;
+    return F;
+  }
+
+  std::optional<FactID> replaceLock(FactManager &FM, const CapabilityExpr &CapE,
+                                    std::unique_ptr<FactEntry> Entry) {
+    return replaceLock(FM, findLockIter(FM, CapE), std::move(Entry));
+  }
+
```
We could, but it's less efficient, esp. if we used findLockIter():
1. without findLockIter: we save a pop_back() and push_back().
2. with preceding findLockIter: like (1), but also saves a search.

Sure, maybe we're just shaving off some constant overheads, but I also find it makes it less verbose and more readable.

I have this later patch I would send out as a separate PR later:
```
commit 9a41f2b68a28e6a2239b45a11a0cd360dce320af
Author: Marco Elver <elver@google.com>
Date:   Fri Apr 25 18:51:16 2025 +0200

    Thread Safety Analysis: Use replaceLock instead of removeLock+addLock

    In ScopedLockableFactEntry::unlock(), we can avoid a second search,
    pop_back(), and push_back() if we use the already obtained iterator into
    the FactSet to replace the old FactEntry and take its position in the
    vector.

diff --git a/clang/lib/Analysis/ThreadSafety.cpp b/clang/lib/Analysis/ThreadSafety.cpp
index cb62f11d6f15..429f22715886 100644
--- a/clang/lib/Analysis/ThreadSafety.cpp
+++ b/clang/lib/Analysis/ThreadSafety.cpp
@@ -1090,9 +1090,9 @@ private:
         return;
       }

-      FSet.removeLock(FactMan, Cp);
-      FSet.addLock(FactMan,
-                   std::make_unique<LockableFactEntry>(!Cp, LK_Exclusive, loc));
+      FSet.replaceLock(
+          FactMan, It,
+          std::make_unique<LockableFactEntry>(!Cp, LK_Exclusive, loc));
     } else if (Handler) {
       SourceLocation PrevLoc;
       if (const FactEntry *Neg = FSet.findLock(FactMan, !Cp))
```

melver replied:
replaceLock without caller-iterator still saves a copy+pop_back+push_back. Not much, but not nothing either.

This PR should include the bare minimum use of replaceLock - I have one patch queued that is not part of this PR to do an unrelated substitution with replaceLock.

If anything I'd move replaceLock introduction before this PR, but I think we're just splitting hairs at that point.

aaronpuchert replied:
Well, I guess we can clean this up later.

melver commented on clang/lib/Analysis/ThreadSafetyCommon.cpp:
```diff
@@ -81,26 +81,25 @@ static bool isCalleeArrow(const Expr *E) {
-}
-
-static StringRef ClassifyDiagnostic(QualType VDT) {
+static CapabilityExpr makeCapabilityExpr(const til::SExpr *E, QualType VDT,
```
I'm simplifying this. The only place where it's not yet know if it's a capability is the ScopedLockableAttr case. If we solve that, then we can have makeCapabilityExpr() - or rather, just a constructor that takes QualType.

melver commented on clang/include/clang/Analysis/Analyses/ThreadSafetyCommon.h:
```diff
@@ -388,7 +395,7 @@ class SExprBuilder {
 
   // Create placeholder for this: we don't know the VarDecl on construction yet.
-  std::pair<til::LiteralPtr *, StringRef>
+  std::pair<til::LiteralPtr *, CapabilityExpr>
```
I think this code is just more complex than it needs to be. I do not understand the need for createThisPlaceholder(): 

- Placeholder.second is only used if there's a ScopedLockableAttr;
- Placeholder.first is equivalent to a createVariable(nullptr);

Therefore, we can just defer the second step (getting the string+bool out) for when we determine it's a ScopedLockableAttr, and use createVariable(nullptr) instead of createThisPlaceholder(). The main problem was that the function that extracts the string (and now also the reentrant bool) wasn't available here.

So it'll just be much clearer if we have a constructor that does that for us, and just use that rather than through this complex indirection of "prefetching" the string+bool.

Also saves a few cycles if we simplify it. Let me try that.

melver commented:
> Two more things come to mind:
> 
> * Consider adding a check to `SemaDeclAttr.cpp` that the new attribute is always accompanied by `capability` or `lockable`. Although I wonder whether that's too early. I'm not sure if we can see the other attributes already.

Added.

> * You should probably do something in `ThreadSafetyAnalyzer::intersectAndWarn`: if two branches join with different counts, we should warn and drop the fact with higher count.

Fixed.

Also addressed all other comments as best as I could.

PTAL.

aaronpuchert commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -876,14 +907,25 @@ class LockableFactEntry : public FactEntry {
                     const CapabilityExpr &Cp, SourceLocation UnlockLoc,
                     bool FullyRemove,
                     ThreadSafetyHandler &Handler) const override {
+    if (std::unique_ptr<FactEntry> RFact = leaveReentrant()) {
+      // This capability remains reentrantly acquired.
+      FSet.replaceLock(FactMan, Cp, std::move(RFact));
+      return;
+    }
+
     FSet.removeLock(FactMan, Cp);
     if (!Cp.negative()) {
       FSet.addLock(FactMan, std::make_unique<LockableFactEntry>(
```
Interesting that both branches remove one lock and add another. I wonder if we shouldn't just do `FSet.removeLock(FactMan, Cp);` unconditionally and then add the other lock. (Side note: I don't know why we have `if (!Cp.negative())`, it seems weird to unlock a negative capability.)

The same would be true for `handleLock`, if we didn't remove the negative capability in `ThreadSafetyAnalyzer::addLock` instead.

melver replied:
Simplifying this.

aaronpuchert commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -894,6 +936,30 @@ class LockableFactEntry : public FactEntry {
+
+  std::unique_ptr<FactEntry> leaveReentrant() const override {
+    if (!ReentrancyDepth)
+      return nullptr;
```
After this I would add `assert(reentrant());`.

aaronpuchert commented on clang/include/clang/Analysis/Analyses/ThreadSafetyCommon.h:
```diff
@@ -271,26 +272,34 @@ class CFGWalker {
+  CapabilityExpr(const til::SExpr *E, StringRef Kind, bool Neg, bool Reentrant)
+      : CapExpr(E, (Neg ? FlagNegative : 0) | (Reentrant ? FlagReentrant : 0)),
+        CapKind(Kind) {}
+  CapabilityExpr(const til::SExpr *E, QualType QT, bool Neg);
```
Ok, I see why you did it this way. It allows you to get rid of `createThisPlaceholder`. I'm not sure if replacing this by `createVariable(nullptr)` doesn't go too deep into TIL implementation details, but I guess it can be justified.

It's just weird that `translateAttrExpr` translates the Clang expression, and the constructor translates the Clang type. But maybe I'll have to live with that.

melver replied:
It's a bit of a maze, but I felt this is a decent trade-off. Less/simpler code traded off against (minimal) loss of abstraction.

aaronpuchert commented on clang/lib/Analysis/ThreadSafetyCommon.cpp:
```diff
@@ -1011,6 +979,30 @@ void SExprBuilder::exitCFG(const CFGBlock *Last) {
   IncompleteArgs.clear();
 }
 
+static CapabilityExpr makeCapabilityExpr(const til::SExpr *E, QualType VDT,
+                                         bool Neg) {
+  // We need to look at the declaration of the type of the value to determine
+  // which it is. The type should either be a record or a typedef, or a pointer
+  // or reference thereof.
+  if (const auto *RT = VDT->getAs<RecordType>()) {
+    if (const auto *RD = RT->getDecl())
+      if (const auto *CA = RD->getAttr<CapabilityAttr>())
+        return CapabilityExpr(E, CA->getName(), Neg,
+                              RD->hasAttr<ReentrantCapabilityAttr>());
+  } else if (const auto *TT = VDT->getAs<TypedefType>()) {
+    if (const auto *TD = TT->getDecl())
+      if (const auto *CA = TD->getAttr<CapabilityAttr>())
+        return CapabilityExpr(E, CA->getName(), Neg,
+                              TD->hasAttr<ReentrantCapabilityAttr>());
+  } else if (VDT->isPointerOrReferenceType())
+    return makeCapabilityExpr(E, VDT->getPointeeType(), Neg);
+
+  return CapabilityExpr(E, StringRef("mutex"), Neg, false);
+}
```
Forgot to mention: the reason why I'd prefer keeping a "classify" function is that you don't have to plumb all the other things (`E`, `Neg`) through it. I think this makes it clearer.

melver replied:
I'll rework this, trying to improve readability here as well.

aaronpuchert commented on clang/lib/Analysis/ThreadSafetyCommon.cpp:
```diff
@@ -1011,6 +979,30 @@ void SExprBuilder::exitCFG(const CFGBlock *Last) {
+  return CapabilityExpr(E, StringRef("mutex"), Neg, false);
+}
+
+CapabilityExpr::CapabilityExpr(const til::SExpr *E, QualType QT, bool Neg)
+    : CapabilityExpr(makeCapabilityExpr(E, QT, Neg)) {}
```
In the header, `CapabilityExpr` seems to come before `SExprBuilder`, so this should probably also come before `SExprBuilder` functions. Or at least before `translateAttrExpr`, so that we can leave the classification in place.

melver commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -4048,6 +4048,9 @@ def warn_thread_attribute_not_on_scoped_lockable_param : Warning<
   "reference to a 'scoped_lockable'-annotated type">,
   InGroup<ThreadSafetyAttributes>, DefaultIgnore;
+def warn_reentrant_capability_without_capability : Warning<
+  "ignoring %0 attribute on %1 without 'capability' attribute">,
```
I'll generalize the diagnostic and make both parameters, so it can be reused if necessary. I think "requires" is not the right word here, because of the new ordering constraint, so I'll choose "must be preceded".

```
--- a/clang/include/clang/Basic/DiagnosticSemaKinds.td
+++ b/clang/include/clang/Basic/DiagnosticSemaKinds.td
@@ -4048,8 +4048,8 @@ def warn_thread_attribute_not_on_scoped_lockable_param : Warning<
   "%0 attribute applies to function parameters only if their type is a "
   "reference to a 'scoped_lockable'-annotated type">,
   InGroup<ThreadSafetyAttributes>, DefaultIgnore;
-def warn_reentrant_capability_without_capability : Warning<
-  "%0 attribute on %1 must be preceded by 'capability' attribute">,
+def warn_thread_attribute_requires_preceded : Warning<
+  "%0 attribute on %1 must be preceded by %2 attribute">,
   InGroup<ThreadSafetyAttributes>, DefaultIgnore;
 def err_attribute_argument_out_of_bounds_extra_info : Error<
   "%0 attribute parameter %1 is out of bounds: "
diff --git a/clang/lib/Sema/SemaDeclAttr.cpp b/clang/lib/Sema/SemaDeclAttr.cpp
index 68ddd286bce5..93d919102a9c 100644
--- a/clang/lib/Sema/SemaDeclAttr.cpp
+++ b/clang/lib/Sema/SemaDeclAttr.cpp
@@ -6249,8 +6249,8 @@ static void handleReentrantCapabilityAttr(Sema &S, Decl *D,
   // This helps enforce a canonical style. Also avoids placing an additional
   // branch into ProcessDeclAttributeList().
   if (!D->hasAttr<CapabilityAttr>()) {
-    S.Diag(AL.getLoc(), diag::warn_reentrant_capability_without_capability)
-        << AL << cast<NamedDecl>(D);
+    S.Diag(AL.getLoc(), diag::warn_thread_attribute_requires_preceded)
+        << AL << cast<NamedDecl>(D) << "'capability'";
     return;
   }
```

melver commented on clang/include/clang/Analysis/Analyses/ThreadSafety.h:
```diff
@@ -94,16 +94,17 @@ enum AccessKind {
   LEK_LockedSomePredecessors,
+  /// A capability is still locked at the end of a function.
   LEK_LockedAtEndOfFunction,
+  /// A capability is locked with mismatching reentrancy depth in predecessors
+  /// of a CFGBlock.
+  LEK_LockedReentrancyMismatch,
```
Using a bool, and a `%select` on the existing warnings. That also produces a better message for loop warnings now.

melver commented:
Thanks for the feedback. Addressed comments as best as I could.

Most notable changes:
- Also warns properly for loops with mismatching reentrancy depth.
- Devirtualized new helpers.
- Require ordering `reentrant_capability` after `capability`.
- Stylistic improvements.

PTAL.

melver commented:
Gentle ping - PTAL.

Many thanks!

melver commented:
As additional motivation - quote from a kernel maintainer:

> But I think we should get the infrastructure in once your reentrancy
> support has landed in a release, because with that we can start
> annotation some code and show uses, while also helping to driver more
> requirements.

So there's hope we don't have to also get alias analysis before the Linux kernel use can proceed.

aaronpuchert reviewed (approved):
I think this looks good, thanks for your contribution!
